src/java.base/share/classes/jdk/crac/impl/OpenFDPolicies.java

```diff
@@ -0,0 +1,307 @@
+package jdk.crac.impl;
+
+import jdk.crac.Context;
+import jdk.crac.Resource;
+import jdk.internal.crac.Core;
+import jdk.internal.crac.JDKResource;
+import jdk.internal.crac.JDKResourceStub;
+import jdk.internal.crac.LoggerContainer;
+import sun.security.action.GetPropertyAction;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.PathMatcher;
+import java.util.*;
+import java.util.regex.Pattern;
+
+import static jdk.crac.impl.OpenFDPolicies.AfterRestore.OPEN_OTHER;
+
+public class OpenFDPolicies<P> {
+    public static final String CHECKPOINT_PROPERTY = "jdk.crac.fd-policy.checkpoint";
+    public static final String RESTORE_PROPERTY = "jdk.crac.fd-policy.restore";
```

Comment:
Having separated policies for checkpoint and restore enables some weird configuration, when e.g. checkpoint specifing CLOSE and restore -- REOPEN. It would be better to have a combined, consisted checkpoint-restore policy that specifes both parts at once.

Reply:
I've started with a single policy enum but it turned out the inlined cross product of behaviours was rather a long and repetitive.
In fact CLOSE + REOPEN is the combination that makes a perfect sense. You close the file on checkpoint (rather than error out and fail the checkpoint) and then you reopen the same file (rather than opening something else).

Reply:
And note that the restore configuration should be set on restore, because in case of opening other files you don't know where these are in the deployment (restore) environment during checkpoint. Had you specified the policy together you would be overriding the behaviour on checkpoint which was already executed, which is even more confusing.

Reply:
> I've started with a single policy enum but it turned out the inlined cross product of behaviours was rather a long and repetitive.

Do you have artifacts of that? Because it may mean that cross-product may have a few invalid / unsafe behaviors. It would be interesting to look at the list.

> In fact CLOSE + REOPEN is the combination that makes a perfect sense. 

Indeed, bad example. What about ERROR+OPEN_OTHER. Or for the sockets, for the current implementation we should not allow OPEN_OTHER, as the implementation can only throw RestoreException in this case.

Reply:
> And note that the restore configuration should be set on restore, because in case of opening other files you don't know where these are in the deployment (restore) environment during checkpoint. Had you specified the policy together you would be overriding the behaviour on checkpoint which was already executed, which is even more confusing.

The combined policy can have additional parameters settable on restore. But the combined policy should be able to limit policies or parameters for restore part of the policy.

Reply:
When the behaviour is ERROR, you won't have a chance to use OPEN_OTHER, because you never get to restore.

If you set policy OPEN_OTHER on FD 42, the restore will throw an exception. We could completely fail the restore and exit, too. So yes, we do not allow that.
We could fail when already reading policies that look like SOCKET=OPEN_OTHER:..., why not add an extra check. But failing when we actually try to do that is only a fraction of second later.